nixos/modules/services/misc/matrix-appservices/default.nix

```diff
@@ -0,0 +1,150 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.matrix-appservices;
+  asOpts = import ./as-options.nix {
+    inherit lib pkgs;
+    systemConfig = config;
+  };
+  mkService = name: opts:
+    with opts;
+    let
+      settingsFormat = pkgs.formats.json { };
+      dataDir = "/var/lib/matrix-as-${name}";
+      registrationFile = "${dataDir}/${name}-registration.yaml";
+      # Replace all references to $DIR to the dat directory
+      settingsData = settingsFormat.generate "config.json" settings;
+      settingsFile = "${dataDir}/config.json";
+      serviceDeps = [ "network-online.target" ] ++ serviceDependencies;
+    in
+    {
+      description = "A matrix appservice for ${name}.";
+
+      wantedBy = [ "multi-user.target" ];
+      wants = serviceDeps;
+      after = serviceDeps;
+      # Appservices don't need synapse up, but synapse exists if registration files are missing
+      before = mkIf (cfg.homeserver != null) [ "${cfg.homeserver}.server" ];
+
+      environment = {
+        DIR = dataDir;
+        SETTINGS_FILE = settingsFile;
+        REGISTRATION_FILE = registrationFile;
+      };
+
+      preStart = ''
+        ${pkgs.envsubst}/bin/envsubst \
+          -i ${settingsData} \
+          -o ${settingsFile}
+        chmod 640 ${settingsFile}
+
+        ${optionalString (registerScript != "") ''
+          if [ ! -f ${registrationFile} ]; then
+            ${registerScript}
+            chmod 640 ${registrationFile}
+          fi
+        ''}
+      '';
+
+      script = startupScript;
+
+      serviceConfig = {
+        Type = "simple";
+        Restart = "always";
+
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        ProtectKernelTunables = true;
+        ProtectKernelModules = true;
+        ProtectControlGroups = true;
+
+        User = "matrix-appservice";
```

Comment:
For all my modules I've been using `DynamicUser`, seems nicer to me to have each appservice running more in its own little world.

Reply:
I wanted to use DynamicUser originally. The main issue is permissions. It got really complicated to ensure synapse gets access to registration files for each appservice, since they would all get their own user. And systemd has a lot of rules for DynamicUser that makes this even more difficult - you can't just add synapse user to the relevant group or add the Dynamic user to synapse's group.

One user for all of them is much easier to setup permissions and it allows them all to store data in one folder. With this method, you can add the appservice, and the registration files are seamlessly created and added to synapse in one rebuild.

Reply:
Definitely a good idea to have them keep their own registration up to date, can confirm that copying that over manually is not cool.

I think for `DynamicUser` one can use `SupplementaryGroups`.

All they need to do is share a directory which is: readable by synapse, writable by the dynamic user, or one of its supplementary groups. I would assume that this is a pattern well-supported by systemd, although I don't quite know the details of how.

Reply:
I've successfully used `DynamicUser` and `SupplementaryGroups` in the past, I'd recommend trying it out for this.

Reply:
A couple other issues I encountered with DynamicUser:
 - the state directory doesn't exist unless the appservice is running and that makes the whole setup less portable, synapse needs the registration files to startup, so if one appservice stops then synapse starts up it will also crash. I think this is because of how systemd implements DynamicUser
 - I was hoping to get all appservices to store data in one folder in var/lib, but thats not possible if they each get their own user.

I will see if I can experiment with SupplementaryGroups, maybe that can help with the permissions problems. I guess I could get the registration to happen with matrix-synapse.

Reply:
Pros/cons. I think that is a case that one has to deal with at that point.
The Matrix server can of course just require the bridges before it starts, and then it should be good-to-go I suppose :)

Reply:
So I did some experimenting and from what I can tell nothing can access the StateDirectory except the one service when DynamicUser is set. It seems to be because systemd stores DynamicUser state data in /var/lib/private and handles permissions while the service is running. So synapse can't access registration files inside those directories at all whether the appservice is running or not. From what I remember permissions wasn't a big deal at all, its just the DynamicUser implementation in systemd

I think I'm going too leave the implementation as it is because of the permissions and data storage headache when dealing with DynamicUser. I see the security of advantage of preventing appservices from accessing other appservice's sensitive information. But I don't know how to get DynamicUser to work without requiring a user to manually copy things. Maybe we can use systemd's InaccessiblePaths to get the same effect.

But if anyone has any ideas to get around this, let me know or feel free to send a PR to this branch.

Reply:
Yeah, for sure trying to pass these files using the appservice's state directory isn't going to work. There needs to be some independent directory where the appservices can dump their registrations and synapse can read them. I would suggest that this directory be writable by an appservice group (also mentioned in the appservice's supplemtarygroups) and readable by synapse itself.

Reply:
I have done something similar in my own PR, that this superseded :)

https://github.com/NixOS/nixpkgs/pull/110404/files#diff-fa863c4551341a3c7ed8d710febc1af635894ccf8f7e8f83d57a69075fc1dc2cR199-R210

Reply:
I would think the main goal of DynamicUser is to prevent each appservice from reading another's file. Wouldn't a directory containing all registration files - which all appservices can access - defeat the point of DynamicUser?

Reply:
being able to read just the registration files is probably better than having full access to each others state, but I'd have thought that it'd be possible to have a write-only directory from the perspective of these services. An alternative would be having some other privileged service keeping these in sync, much like users might manually copy over registration files for the current state of appservices.

As much as I hate to belabour the discussion on this excellent PR, these are quite important considerations from a security perspective given that these services shuttle around people private conversations.

Reply:
Ahh I see your right we are worried about the case of, for example, whatsapp accessing the database of the discord appservice. Which could pose a security issue. 

So then instead of a separate folder or getting another service to manage it, could we make use of `InaccessiblePath`? We could forbid each appservice from accessing any other appservice's state directory but keep them running on the same user. That way permissions can be controlled in a much cleaner way - I prefer the one user/group for permission control. And matrix-synapse is the one service we always trust to get access to all appservice's paths.

Reply:
I suppose that could work, but it just seems more complicated and error prone (what if we don't cover all the used paths). 

I think that the simplest solution is just to have a directory writable (perhaps with `BindPaths`) by the bridges and readable by synapse, having the bridges put the registrations in there and chown them to synapse (or a relevent group).

Reply:
I think this problem is solved with Pacman99/nixpkgs#1